buffer/src/main/java/io/netty/buffer/TestingPooledByteBufAllocator.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2018 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.buffer;
+
+import io.netty.util.ResourceLeakDetector;
+import io.netty.util.ResourceLeakTracker;
+
+import java.io.Closeable;
+
+/**
+ * A {@link ByteBufAllocator} for unit testing. This allocator maintains a private, non-static leak detector
+ * and performs leak tracking on all allocated buffers. At the end the test, call {@link #close} to assert
+ * that all buffers have been freed.
+ */
+public class TestingPooledByteBufAllocator extends PooledByteBufAllocator implements Closeable {
```

Comment:
As this is mainly used for testing I think we should better extend `UnpooledByteBufAllocator`.

Reply:
Also maybe we should call it `AssertingByteBufAllocator`.

Reply:
@normanmaurer, what are the implications of extending `UnpooledByteBufAllocator`?  I ask because I always use pooled in my tests because that is what I use in production.  IIRC, the unpooled uses different methods for the leak callbacks.